### PR TITLE
ci: [M12] CONF-2: cross-backend MLX↔CUDA parity is never actually checked by CI

### LINKED ISSUE #303: [M12] CONF-2: cross-backend MLX↔CUDA parity is never actually checked by CI
Found by `/closure-audit` (2026-08-18), pass C. Capability row `CONF-2` in
`docs/feature-matrix.md`.

## Capability
A contributor can trust that a change keeps the MLX and CUDA backends numerically identical,
because CI proves it on every change — rather than because someone once ran it by hand.

## The gap

`tests/test_backend_parity.py` is well-written and guards the thing `CLAUDE.md` calls a core
invariant. It also **never runs**. Five of its six tests are guarded by
`@pytest.mark.skipif(not (HAVE_MLX and HAVE_TORCH))`:

| line | test |
|---|---|
| `:48` | `test_backend_parity_mlx_vs_torch` |
| `:76` | `test_backend_parity_hybrid` |
| `:103` | `test_backend_parity_seg_ids` |
| `:137` | `test_portable_weights_roundtrip_both_directions` |
| `:168` | `test_moe_routing_entropy_parity_mlx_vs_torch` |

No job in `.github/workflows/ci.yml` satisfies that condition:

- mlx jobs — `full-macos` (`.[dev,data,mlx]`), `swift-engine`, `poc-fixture-oracle`,
  `poc-parity` (`.[dev,mlx]`) — install **no torch**.
- torch jobs — `smoke-linux`, `cuda-cpu` — run on `ubuntu-latest`, where **mlx is not
  installable at all**.

So `HAVE_MLX and HAVE_TORCH` is false in every job, every run. The only test in the file that
executes in CI is `test_parity_harness_torch_self` at `:205`, which compares torch against
itself. The file's own docstring says the real comparisons are "runnable entirely on a Mac (mlx +
torch both present)" — that Mac is a developer laptop, not CI.

This is the failure shape the repo's own standing rules warn about: a green suite and a check
that cannot observe its target are indistinguishable from the outside.

## Scope

In scope:
- Install CPU torch alongside mlx in the existing `full-macos` job (`macos-latest` supports both),
  or add one small macOS job that has both, whichever keeps `full-macos`'s runtime acceptable.
- Make the five skipped tests execute there.
- Make the skip **loud**: if the job that is supposed to have both backends does not, fail rather
  than skip, so this cannot silently regress to the current state.

Out of scope:
- Real-GPU CUDA parity (no GPU runner; `cuda-cpu` semantics are what is being compared).
- Widening the tolerance to make a failing comparison pass — if the backends genuinely disagree
  at `~1e-4` fp32, that is a separate bug and gets its own issue.

## Definition of done
- [ ] A CI job installs both `mlx` and CPU `torch`, and the five tests at `:48,76,103,137,168`
      **run** there — verified by reading the job's pytest output, not by the job being green.
- [ ] That job reports the five as passed, not skipped. `pytest -rs` output is checked for them.
- [x] A guard makes a future silent regression impossible: in the designated both-backends job,
      absent mlx or torch is an **error**, not a skip.
- [ ] `full-macos` runtime stays acceptable, or the work moves to its own job.
- [x] `docs/test-plan.md`'s `CONF-2` row is updated — the Gaps cell currently states this gap
      explicitly and must be rewritten to match reality.
- [x] `docs/design/03-conformance.md` reflects that backend parity is now automated, replacing the
      "skip cleanly on a single-backend host" framing as the *only* story.
- [ ] Full suite still green on every existing job; no unrelated behaviour changed.

## Edge cases
- CPU torch and mlx sharing a macOS runner: version conflicts, or torch pulling a numpy that mlx
  disagrees with. Pin what is needed.
- The comparison may **fail on first run** — it has genuinely never executed in CI. That is a
  finding, not a reason to loosen the tolerance.
- MoE routing-entropy parity (`:168`) is the most likely to be tolerance-sensitive; treat a
  failure there as a real result and investigate before adjusting.
- `test_portable_weights_roundtrip_both_directions` (`:137`) writes checkpoints — confirm the
  runner has the disk and that temp dirs are cleaned.

## Verification
- `.venv/bin/python -m pytest tests/test_backend_parity.py -q -rs` on a host with both backends —
  expect **6 passed, 0 skipped**.
- In CI: the job log for `tests/test_backend_parity.py` shows five passes and no `s` markers.
- `scripts/smoke_test.py` on both backends unchanged.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #303

## Summary

Five of the six tests in `tests/test_backend_parity.py` — the real MLX-vs-CUDA/torch
comparisons — **skipped in every CI job** for the whole M12 build. `portable` has neither
backend, `smoke-linux`/`cuda-cpu` have torch only, `full-macos` had mlx only. The only
cross-backend test that had ever executed was `test_parity_harness_torch_self`, which
compares torch against *itself*. A check that cannot observe its target is
indistinguishable from the good outcome.

**One runner can host both.** The macOS-arm64 PyPI `torch` wheel is CPU-only (there is no
CUDA build for macOS), and CPU torch is exactly the surface `src/model/cuda_backend.py` is
compared on — `pyproject.toml`'s `cuda` extra already says as much. So no artifact
handoff and **no new job**:

- `.github/workflows/ci.yml` — `full-macos` installs `.[dev,data,mlx,cuda]` (plain PyPI,
  *not* the torch CPU index, which carries no macos-arm64 wheels) and gains a dedicated
  step carrying `MONICA_REQUIRE_BOTH_BACKENDS: "1"` that runs `tests/test_backend_parity.py`.
  The main `pytest -q -rs` step deliberately does **not** carry the flag, so the rest of
  the suite keeps normal skip semantics.
- `tests/test_backend_parity.py` — the five literal `skipif`s become a shared
  `requires_both_backends` decorator that, under the flag, attaches **no marker at all**.
  Plus two guards (below).
- `tests/test_ci_backend_matrix.py` (new) — the workflow-YAML contract. Portable
  (`yaml` + stdlib), so it runs on the Linux `portable` job.

**No `src/` file is touched and no tolerance is changed** — the comparison stays fp32
`rtol=1e-4 / atol=1e-5`. `tests/test_workflow_triggers.py` is **unedited** and passes,
which is the proof no job was added, removed, or moved between the two workflow files.

### Four guard layers

1. `requires_both_backends` returns the function unwrapped under the flag → a missing
   backend raises `ImportError` from the test body. `test_designated_job_has_both_backends`
   says the same thing up front with a message naming the cause.
2. `test_cross_backend_tests_carry_no_skip_marker_when_required` fails if any of the five
   regains a `skip`/`skipif`. `globals()[name]` raises `KeyError` on a rename — the tuple
   of five is a contract, and a rename that drops one from the gate is the regression.
3. `tests/test_ci_backend_matrix.py` asserts exactly one job declares the flag, that it is
   `full-macos`, that it is set to the literal `"1"`, that the job runs on `macos-*`, that
   it installs both extras (parsing the extras list, not substring-matching the script),
   and that the flagged step actually runs the parity file.
4. pytest's exit code 5 if the file is deleted or renamed away.

### The negative control was actually run and observed to fail

An unfired guard is an unproven guard. With `mlx` hidden behind a shim package on
`PYTHONPATH`:

```
$ PYTHONPATH=$SHIM .venv/bin/python -m pytest -q -rs tests/test_backend_parity.py
sssss.ss    1 passed, 7 skipped        # today's behaviour: silent skip

$ MONICA_REQUIRE_BOTH_BACKENDS=1 PYTHONPATH=$SHIM .venv/bin/python -m pytest -q -rs \
      tests/test_backend_parity.py
FFFFF.F.    6 failed, 2 passed         # zero skips
E  AssertionError: MONICA_REQUIRE_BOTH_BACKENDS=1 but `import mlx.core` failed — this job
   is the cross-backend parity gate and its install step must resolve an mlx wheel
```

`FFFFF.F.` reads: the five comparisons **error**, the torch self-check still passes, the
designated-job guard fails, the no-skip-marker guard passes. **No `s`.**

Both `ci.yml` mutations were also observed red and reverted:

| Hand-mutation | Result |
|---|---|
| delete the `env: MONICA_REQUIRE_BOTH_BACKENDS: "1"` block | 3 failed — `ci.yml: no step of full-macos carries MONICA_REQUIRE_BOTH_BACKENDS` |
| `.[dev,data,mlx,cuda]` → `.[dev,data,mlx]` | 1 failed — `full-macos no longer installs ['cuda']` |

(The first attempt at the second mutation was a *bad mutation* — `perl -0p` without `/g`
hit the banner comment instead of the install line, so the test correctly stayed green.
The extras check was then tightened to parse the `pip install -e ".[...]"` extras list so
a comment mentioning `cuda` cannot satisfy the contract, and re-verified red.)

## What is *not* proven, stated in the docs

The `CONF-2` Gaps cell is **not** `none`. The gate covers **toy scale, fp32 only**
(`B=2, L=24`) against **torch on CPU**; real-GPU CUDA kernels and poc-scale shapes are
compared by **no** job. `cuda-cpu` (Linux) still skips 5 of 6 — mlx has no Linux wheel, so
`cuda-cpu` green is **not** parity evidence. All three statements are now in
`docs/design/03-conformance.md` and `docs/test-plan.md`.

## Docs updated in the same commit as the code

- `docs/feature-matrix.md` — `CONF-2` `In progress` → `Shipped`
- `docs/test-plan.md` — `CONF-2` row: E2E `none` → `CI full-macos`; edge cases name all
  five comparisons; Gaps rewritten per above
- `docs/design/03-conformance.md` — `## Status` rewritten; the `backend-vs-backend`
  section now points at it
- `CLAUDE.md` — `full-macos` described as carrying both backends and as the cross-backend
  parity gate

## Testing

- [x] Tests added/updated — `tests/test_ci_backend_matrix.py` (5 tests) + 2 guards in
      `tests/test_backend_parity.py`
- [x] All tests passing — `.venv/bin/python -m pytest -q -rs` → **1671 passed, 14 skipped,
      6m13s** (the plan's 1666/12 baseline plus exactly the 5 new + 2 new-skipping tests)
- [x] `MONICA_REQUIRE_BOTH_BACKENDS=1 pytest -q -rs tests/test_backend_parity.py` →
      **8 passed, 0 skipped** — reproduces the CI job's semantics exactly
- [x] `pytest -q tests/test_workflow_triggers.py` → passes, **unedited**
- [x] M4 smoke gate on a fresh toy split → `SMOKE TEST PASSED ✅`, resume diff 1.192e-07
- [x] Negative control observed to fail (above)

## To check on this PR's CI run

- **Criterion 2 is read from the log text, not the checkmark**: `full-macos` → step
  *"Cross-backend parity"* must show `8 passed` with **no** `SKIPPED` line naming any of
  the five.
- **Edge case E2**: `tests/test_cuda_compile.py` newly executes on macOS. It skips locally
  (this machine's clang cannot find `<algorithm>`), but a GitHub macOS runner has full
  Xcode CLT, so inductor may genuinely compile there — minutes, possibly flaky. The
  `_compile_works` fixture already skips on compile failure. If it fails for a
  macOS-specific reason, widen that fixture; **do not** widen any parity tolerance.
- **Criterion 5**: `full-macos` wall clock ≤ ~15 min (baseline 9m16s, `timeout-minutes: 45`).
- **E3/E4**: a cross-backend failure on the runner is a **real finding** — report the
  deviation and open a separate issue. Tolerance is explicitly out of scope.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_011ur4BdaFTqnwqSUAx5TrX8

### CODE REVIEW

File: tests/test_ci_backend_matrix.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    """The step carrying the flag must actually run the file it unlocks. A flag set on a
+    step that runs something else is a gate pointed at nothing."""
+    job = _ci_jobs()[DESIGNATED_JOB]
+    flagged = [step for step in job.get("steps") or [] if FLAG in (step.get("env") or {})]
+    assert flagged, f"{CI}: no step of {DESIGNATED_JOB} carries {FLAG}"
+    for step in flagged:
+        assert "tests/test_backend_parity.py" in str(step.get("run", "")), (
+            f"{CI}: the step carrying {FLAG} does not run tests/test_backend_parity.py; "
+            f"the flag would be set for a command it does not affect"
+        )
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply tests/test_ci_backend_matrix.py:175]: ✅ Fixed in a68fb79 — `test_the_parity_step_targets_the_parity_test_file` now accepts a job-level `MONICA_REQUIRE_BOTH_BACKENDS` (requiring only that some step in the job runs the parity file) as well as step-level (requiring every flagged step to run it), matching what `_declaring_jobs`/`test_exactly_one_job_declares_the_both_backends_flag` already allow.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -149,13 +156,27 @@ jobs:
...
+        # tests/test_backend_parity.py's `requires_both_backends`), so a missing backend
+        # ERRORS instead of skipping; pytest's exit code 5 ("no tests collected") catches
+        # the file being deleted or renamed away. Expect `8 passed`, 0 skipped.
+        env:
+          MONICA_REQUIRE_BOTH_BACKENDS: "1"
+        run: pytest -q -rs tests/test_backend_parity.py
       - name: pytest
+        # No MONICA_REQUIRE_BOTH_BACKENDS here on purpose: the rest of the suite keeps
+        # its normal skip semantics.
         run: pytest -q -rs
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply .github/workflows/ci.yml]: ✅ Fixed in a68fb79 — the full-suite `pytest` step in `full-macos` now runs with `--ignore=tests/test_backend_parity.py`, since the dedicated Cross-backend parity step already runs that file under `MONICA_REQUIRE_BOTH_BACKENDS=1`.

**Comment by travisgalloway** [comment]
Addressed review feedback:

- @copilot-pull-request-reviewer: overview only, no items beyond the 2 inline comments below — both fixed in a68fb79.
